File: internal/usecases/dto/memo.go
package dto

import "strconv"

// =====================
// MemoForm は質問メモのフォーム DTO である。
type MemoForm struct {
	ID         int64         `json:"id"`
	QuestionID string        `json:"questionId"`
	UserID     string        `json:"userId"`
	Content    string        `json:"content"`
	CreatedAt  *string       `json:"createdAt,omitempty"`
	UpdatedAt  *string       `json:"updatedAt,omitempty"`
	DeletedAt  *string       `json:"deletedAt,omitempty"`
	Question   *QuestionForm `json:"question,omitempty"`
	User       *UserForm     `json:"user,omitempty"`
}

// QuestionIDInt64 は QuestionID 文字列を int64 に変換する。
//
// return:
//   - int64: 質問 ID（変換失敗時は -1）
func (f MemoForm) QuestionIDInt64() int64 {
	if val, err := strconv.ParseInt(f.QuestionID, 10, 64); err == nil {
		return val
	}
	return -1
}

// UserIDInt64 は UserID 文字列を int64 に変換する。
//
// return:
//   - int64: ユーザー ID（変換失敗時は -1）
func (f MemoForm) UserIDInt64() int64 {
	if val, err := strconv.ParseInt(f.UserID, 10, 64); err == nil {
		return val
	}
	return -1
}

File: static/js/models/Memo.js
/**
 * Memo model class.
 */

import { Question } from '/static/js/models/Question.js';
import { User } from '/static/js/models/User.js';
import { formToApiJson } from '/static/js/models/helpers.js';

export class Memo {
  /**
   * @param {number}       id
   * @param {number}       questionId
   * @param {number}       userId
   * @param {string}       content
   * @param {Question|null} [question]
   * @param {User|null}    [user]
   * @param {string|null} createdAt
   * @param {string|null} updatedAt
   */
  constructor(id, questionId, userId, content, question = null, user = null, createdAt = null, updatedAt = null) {
    this.id = id;
    this.questionId = questionId;
    this.userId = userId;
    this.content = content;
    this.question = question;
    this.user = user;
    this.createdAt = createdAt ? new Date(createdAt) : null;
    this.updatedAt = updatedAt ? new Date(updatedAt) : null;
  }

  /** @param {Object} json @returns {Memo} */
  static fromJSON(json) {
    return new Memo(
      json.id,
      json.questionId,
      json.userId,
      json.content,
      json.question ? Question.fromJSON(json.question) : null,
      json.user ? User.fromJSON(json.user) : null,
      json.createdAt ?? null,
      json.updatedAt ?? null,
    );
  }

  /** @returns {Object} */
  static toModel(form) {
    return formToApiJson(form);
  }
}
